platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileCompactor.java

```diff
@@ -0,0 +1,482 @@
+        for (DataFileReader<?> r : filesToCompact) {
+            readers[r.getIndex() - firstIndexInc] = r;
+        }
+
+        boolean allDataItemsProcessed = false;
+        try {
+            final KeyRange keyRange = dataFileCollection.getValidKeyRange();
+            index.forEach((path, dataLocation) -> {
+                if (!keyRange.withinRange(path)) {
+                    return;
+                }
+                final int fileIndex = DataFileCommon.fileIndexFromDataLocation(dataLocation);
+                if ((fileIndex < firstIndexInc) || (fileIndex >= lastIndexExc)) {
+                    return;
+                }
+                final DataFileReader<?> reader = readers[fileIndex - firstIndexInc];
+                if (reader == null) {
+                    return;
+                }
+                final long fileOffset = DataFileCommon.byteOffsetFromDataLocation(dataLocation);
+                // Take the lock. If a snapshot is started in a different thread, this call
+                // will block until the snapshot is done. The current file will be flushed,
+                // and current data file writer and reader will point to a new file
+                snapshotCompactionLock.acquire();
+                try {
+                    final DataFileWriter<?> newFileWriter = currentCompactionWriter.get();
+                    long serializationVersion = reader.getMetadata().getSerializationVersion();
+                    final long newLocation = newFileWriter.writeCopiedDataItem(
+                            serializationVersion, reader.readDataItemBytes(fileOffset));
+                    // update the index
+                    index.putIfEqual(path, dataLocation, newLocation);
+                } catch (final IOException z) {
+                    logger.error(EXCEPTION.getMarker(), "Failed to copy data item {} / {}", fileIndex, fileOffset, z);
+                    throw z;
+                } finally {
+                    snapshotCompactionLock.release();
+                }
+            });
+            allDataItemsProcessed = true;
+        } finally {
+            // Even if the thread is interrupted, make sure the new compacted file is properly closed
+            // and is included to future compactions
+            snapshotCompactionLock.acquire();
+            try {
+                // Finish writing the last file. In rare cases, it may be an empty file
+                finishCurrentCompactionFile();
+                // Clear compaction start time
+                currentCompactionStartTime.set(null);
+                if (allDataItemsProcessed) {
+                    // Close the readers and delete compacted files
+                    dataFileCollection.deleteFiles(filesToCompact);
+                }
+            } finally {
+                snapshotCompactionLock.release();
+            }
+        }
+
+        return newCompactedFiles;
+    }
+
+    // visible for testing
+    int getMinNumberOfFilesToCompact() {
```

Comment:
You have access to the Configuration in the test this should not be needed. If you pass it in as recommended or by using `ConfigurationHolder` in the test as well.